Add support for multi game XCIs (second try)

TSRBerry opened the pull request:
I rebased this branch and fixed an issue that I saw. I mainly need more testing to make sure GTK and Ava are both working fine with these changes.

> [!NOTE]
> I added a new command line argument (`-i` or `--application-id`) for Avalonia and GTK, which allows to launch a specific application from a multi game XCI.
> This argument will be ignored if the specified application is not an XCI or if no application was specified on the command line.

## What needs to be tested

For GTK and Avalonia I'd like to know if the following things work:

- Can every game of a multi-game XCI/NSP be launched?
- Can you extract sections (romfs/exefs) from multi-game XCIs/NSPs?
  - Even with some updates and/or DLCs enabled (including bundled ones)?
- Does the bundled update of an XCI show up in the updater window?
  - Can the bundled update be disabled/enabled?
- Can DLCs and updates be added to multi-game XCIs/NSPs?
- Can DLCs and updates still be added to regular XCIs/NSPs?
- If the bundled update/dlc is selected
  - will it be applied on launch from the GUI?
  - will it also be applied if you load the game from file (`File > Load Application from File`)?
  - and the update/dlc window is opened again are there no duplicated entries?
  - and you switch to the other GUI (either GTK or Ava) does the updater/dlc window still open without duplicate entries or other errors?

---

Fixes #6158
Fixes #6740
Fixes #6761

---

<details>
<summary>Original PR description from #5638</summary>

Huge thank you to @piplup55 for his patience and the large amount of testing he did for me while I was trying to make this work!

This PR refactors a lot of the code used for application loading to use `.cnmt.nca` files.

Before we needed a lot more logic for checks that can now be reduced to simple reads of values in the `cnmt` files.

I wrapped `Cnmt` in a `ContentCollection` class which can also get the referenced `NCAs`, so that should be pretty simple now as well.

---

Eventually we should also replace all `ProgramIndex` logic with reads from `NACP`, but that's for another PR.

---

Fixes #3234
Closes #1852
</details>

piplup55 commented:
# Issues

Games with multiple games on them softlock the ui when loading from the file menu
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/ba785d8f-ccc1-4ca1-beef-5d40270469ad)
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/20227fb6-69e1-4963-b79d-ea7ce3f96f70)

Games show the wrong name when loaded from the file menu
Mania:
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/3873f897-7e23-41d4-8bf5-4559dba0b8c0)
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/d97bac0f-ff27-487f-a367-258ced50d3c8)
SMO:
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/b4486b38-5050-4de2-9cc9-5e51e9b0a8c6)
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/8ca18f65-cd57-46ba-867c-0da5642e6dfb)

DLC Window resetting bundled content on reopen
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/b15f2a0c-eac8-47a8-a550-b58246446f66)
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/395ef26b-e8af-47b3-b460-7ed74ef662a9)

DLC Window crashes GTK
```
00:01:09.402 |E| Application : Unhandled exception caught: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> LibHac.Common.HorizonResultException: ResultFsPartitionSignatureVerificationFailed (2002-4644)
   at LibHac.Common.ThrowHelper.ThrowResult(Result result)
   at LibHac.Result.ThrowIfFailure()
   at Ryujinx.UI.Windows.DlcWindow.AddDlc(String path, Boolean ignoreNotFound) in /mnt/data/Github/Clone/Ryujinx/multi-title-xcis-take2/src/Ryujinx.Gtk3/UI/Windows/DlcWindow.cs:line 166
   at Ryujinx.UI.Windows.DlcWindow..ctor(Builder builder, VirtualFileSystem virtualFileSystem, String applicationId, ApplicationData title) in /mnt/data/Github/Clone/Ryujinx/multi-title-xcis-take2/src/Ryujinx.Gtk3/UI/Windows/DlcWindow.cs:line 82
   at Ryujinx.UI.Windows.DlcWindow..ctor(VirtualFileSystem virtualFileSystem, String titleId, ApplicationData title) in /mnt/data/Github/Clone/Ryujinx/multi-title-xcis-take2/src/Ryujinx.Gtk3/UI/Windows/DlcWindow.cs:line 38
   at Ryujinx.UI.Widgets.GameTableContextMenu.ManageDlc_Clicked(Object sender, EventArgs args) in /mnt/data/Github/Clone/Ryujinx/multi-title-xcis-take2/src/Ryujinx.Gtk3/UI/Widgets/GameTableContextMenu.cs:line 479
   at InvokeStub_EventHandler.Invoke(Object, Span`1)
   at System.Reflection.MethodBaseInvoker.InvokeWithFewArgs(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   --- End of inner exception stack trace ---
   at System.Reflection.MethodBaseInvoker.InvokeWithFewArgs(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at GLib.SignalClosure.MarshalCallback(IntPtr raw_closure, IntPtr return_val, UInt32 n_param_vals, IntPtr param_values, IntPtr invocation_hint, IntPtr marshal_data)
```

# Features that work as intended:tm:

Updates can be applied to multi-xci games/can be enabled or disabled
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/5872b36d-c5e2-4876-849b-a4272571faa2)

Updates and dlc can be applied to non multi-xci games
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/55adcdbd-325c-4f32-9dd2-7f4e341360d7)
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/49932325-89d7-4b5b-bbfd-4551f5c92727)

No duplicate entries in gtk for updates
![image](https://github.com/Ryujinx/Ryujinx/assets/100526773/e55ecd29-7e69-4ada-8f00-d866ec9828b2)

erictbar commented:
Loaded ``ROBOTICS_NOTES DaSH v0 (010039A0117C0000) + ROBOTICS_NOTES ELITE v0 (01002A900EE8A000).xci``
ELITE is the main game and DaSH is the spin-off. In 1.1.1239 only DaSH shows up in game list and loads from file menu.

In 1.1.0+b1bc5b9, both games do show up on my game list, and both load from the list. 

However, when loading from the file menu, ELITE loads but with ``Loading Unknown`` and a generic icon. Though, the game does start.
![image](https://github.com/Ryujinx/Ryujinx/assets/332205/0679db30-ea6b-4337-8b38-309c86fde024)

I do not belive this game has any DLC or updates.

TSRBerry commented:
Thank you @piplup55 and @erictbar I think I addressed all the issues with my latest push. Builds will be ready in about 20 minutes if you want to test again!

TSRBerry commented on src/Ryujinx.HLE/Loaders/Processes/Extensions/NcaExtensions.cs:
```diff
@@ -108,6 +128,56 @@ public static bool IsControl(this Nca nca)
 
+        public static (Nca, Nca) GetUpdateData(this Nca mainNca, VirtualFileSystem fileSystem, IntegrityCheckLevel checkLevel, int programIndex, out string updatePath)
+        {
+            updatePath = "(unknown)";
```
It's just used for logging as far as I can tell. Null should work as well!

erictbar commented:
> Thank you @piplup55 and @erictbar I think I addressed all the issues with my latest push. Builds will be ready in about 20 minutes if you want to test again!

Icon & title fixed when loading from File menu, both games still load from game list.

frafes commented:
XCI files with multiple games are loading fine.

However it seems there is an issue when the game is bundled with multiple DLCs.
Loaded `NAMCOT COLLECTION [010039F010E14000][v0].xci`, the game is bundled with 10 DLCs but only one is loaded. If right click the game > Manage DLC in the GUI, it shows only 1 DLC.

The regular Ryujinx loads fine all 10 DLCs, but the emulator crashes when right click> Manage DLC.

TSRBerry commented:
Thank you, it should hopefully detect all of them now!

OddOttAllen commented:
Everything seems to be working just fine for me! I only have one game to test this with though, and that's Pikmin 1+2. It's working just fine, though. Both games load and play, and their updates apply as well.

Hoping that this gets its reviews and merges soon :)

OddOttAllen commented:
> > Everything seems to be working just fine for me! I only have one game to test this with though, and that's Pikmin 1+2. It's working fine. Both games load and play, and their updates apply as well.
> > Though, I'm having sound issues with 1, and 2's title screen doesn't show up. I cannot tell if this is a bug with this feature or bugs with the games themselves, as these are the only copies of the game I have to test.
> > Hoping for the reviews and merge soon :)
> 
> Id try this myself but have you tried games like Klonoa collection, 3D All Stars, or any other games that are multi launch. On other emu usually problems arised when you launch those games with bundled updates

this is for games that have multiple titles show up on the home screen on the switch though, not "multi launch" games

but i guess if it's relevant, the suspend menu doesn't work in 3D all stars?

piplup55 commented:
3D all-stars is unrelated to this pr, the suspend menu is hit and miss afaik

legolas119 commented:
hi, i'm trying Double Dragon Collection (.xci format) but Ryujinx only show me one of the games included: Double Dragon Advance: is it possibile to show all the games included? How can i try to do that?
![double_ryujin](https://github.com/Ryujinx/Ryujinx/assets/48774386/97884b9f-4f57-4a93-afdd-f630c39c6918)

thanks!

TSRBerry commented:
Were you testing this PR? @legolas119

legolas119 commented:
> 

i'm using the latest Ryujinx version released

GamerzHell9137 commented:
> > 
> 
> i'm using the latest Ryujinx version released

Then test this PR instead?
https://github.com/Ryujinx/Ryujinx/pull/6515#issuecomment-2002200107

legolas119 commented:
> > > 
> > 
> > 
> > i'm using the latest Ryujinx version released
> 
> Then test this PR instead? [#6515 (comment)](https://github.com/Ryujinx/Ryujinx/pull/6515#issuecomment-2002200107)

yeah, i tried that version and all the games are displayed:
https://i.postimg.cc/908R40xX/double-dragon.png

Will it be possibile to implement this feature in the official Ryujinx release?
thanks!

Serfrost commented:
> Will it be possibile to implement this feature in the official Ryujinx release? thanks!

That is the plan. PRs have to be tested before they can be merged into the public build, then reviewers need to check over the PR to make sure that all is good. (This isn't exactly the place for a GitHub 101 though.) When you see that this has been tagged as Merged, it should then be in the main build. For now, you can use this PR build to play your game if you wish.

wzsx150 commented:
> yeah, i tried that version and all the games are displayed: https://i.postimg.cc/908R40xX/double-dragon.png
> 
> Will it be possibile to implement this feature in the official Ryujinx release? thanks!

I don't have a compilation environment. Could you share the compiled win64 file? I would like to test it out. Thank you.

legolas119 commented:
thank you very much for the explanation :)

frafes commented:
Found a new issue when launching any game via CLI, on loading screen the [game artwork is missing and the game name is displayed as unknown](https://github.com/Ryujinx/Ryujinx/pull/6515#issuecomment-2002402827).

Also I didn't find a way to launch an especific game from a multi game XCI via command line.

TSRBerry commented:
> Found a new issue when launching any game via CLI, on loading screen the [game artwork is missing and the game name is displayed as unknown](https://github.com/Ryujinx/Ryujinx/pull/6515#issuecomment-2002402827).

Thank you, it should be fixed with my latest commit!

> Also I didn't find a way to launch an especific game from a multi game XCI via command line.

With the latest commit I added a command line argument (`-i` or `--application-id`) for Avalonia and GTK to launch a specific application from a provided file.
If you don't specify a file to launch on the command line this argument will be ignored.

frafes commented:
> With the latest commit I added a command line argument (`-i` or `--application-id`) for Avalonia and GTK to launch a specific application from a provided file.
> If you don't specify a file to launch on the command line this argument will be ignored.

It's working great!

About the GUI version of Ryujinx: when right-click a game there is an option to create a desktop shortcut for it. However, currently, the shortcut dosen't work for multi xci. Would it be possible to automatically add the new command-line argument + application ID to the shortcut when creating it for those games?

Thank you!

TSRBerry commented:
> Would it be possible to automatically add the new command-line argument + application ID to the shortcut when creating it for those games?

Wow, thank you for catching that I forgot about shortcuts! With the latest commit it should be possible to create specific game shortcuts as well.

frafes commented:
> With the latest commit it should be possible to create specific game shortcuts as well.

Shortcuts are working fine now, Thank you!

I tested a few more games (over 40 games) including multi-games.xci, single-game.xci and game.nsp and they all loaded fine, except 4 of them. These four games are single-game.xci:

- MONSTER HUNTER STORIES 2: WINGS OF RUIN v1.0.1 `[0100E21011446000]`
- Puyo Puyo™ Tetris® v1.0.0 `[010073C001D5E000]`
- Puyo Puyo™ Tetris® 2 v1.0.0 `[010038E011940000]`
- Splatoon 2 v1.0.0 `[01003BC0000A0000]`

All of these games load normally in the latest official version of Ryujinx.

I also noticed that with this PR the title-IDs of these four games are detected incorrectly:

![zzz](https://github.com/Ryujinx/Ryujinx/assets/72279502/c40a0ea6-01b8-4667-a507-c35b7132c4bf)


I don't know if this issue occurs with any other untested game.

Error log: [Ryujinx_1.1.0+c1ce433_2024-04-21_21-20-05.log](https://github.com/Ryujinx/Ryujinx/files/15056466/Ryujinx_1.1.0%2Bc1ce433_2024-04-21_21-20-05.log)

TSRBerry commented:
@frafes I hope this is fixed with my latest commit.

The solution for this is not as pretty as I hoped, but it should work and shouldn't cause any problems.
It's just something that will need more refactoring in the future I guess.

frafes commented:
> I hope this is fixed with my latest commit.

Thank you! All those games are loading fine.


While testing some more games I found a new issue:

When adding a game directory path to Ryujinx, the first game listed in that folder (in alphabetical order) cannot be launched either via command line or shortcut.

1. If there is only one game/file in that folder: it fails to launch via command line or shortcut.

2. If there are multiple games/files in that folder: all games can be launched, with the exception of the first one (listed in alphabetical order), when using the command line or shortcut.

However, if I remove the game directory path from Ryujinx (either in the GUI or by editing the PRConfig.json file), all games launch successfully via the command line or shortcut.

Tested both GTK and Avalonia versions, this issue occurs just with the Avalonia version. The latest official Ryujinx Avalonia version doesn't have this issue.

Log file: [Ryujinx_1.1.0+6eb3a42_2024-04-30_00-08-27.log](https://github.com/Ryujinx/Ryujinx/files/15157628/Ryujinx_1.1.0%2B6eb3a42_2024-04-30_00-08-27.log)

piplup55 commented:
> When adding a game directory path to Ryujinx, the first game listed in that folder (in alphabetical order) cannot be launched either via command line or shortcut.

the log file you provided says it's being used by another process

TSRBerry commented:
> If there is only one game/file in that folder: it fails to launch via command line or shortcut.

Do you get an exception and a stacktrace? If so could you post it?

frafes commented:
> Do you get an exception and a stacktrace? If so could you post it?

No, I don’t get an exception and a stacktrace, sorry. Is there a way I can get it?

TSRBerry commented:
Odd, are you able to reproduce this on the command line, so you get to see everything?

frafes commented:
> Odd, are you able to reproduce this on the command line, so you get to see everything?

It is basically the same that log file.
![cmd_ryujinx_print](https://github.com/Ryujinx/Ryujinx/assets/72279502/609ca524-0962-4e50-b244-111c1af8d61b)
Tested in two different Windows 11 PCs, same result.

TSRBerry commented:
Seems like you are running multiple instances at once, or something else is accessing the files.

Doesn't look like an issue I could address here.

frafes commented:
> Doesn't look like an issue I could address here.

No problem. Thank you so much for all your hard work!

TSRBerry commented:
Does that happen on master as well?

frafes commented:
> Does that happen on master as well?

It doesn't happen on master.

TSRBerry commented:
Finally, it should be fixed now!

Anything else that came up during testing? :D

frafes commented:
> Finally, it should be fixed now!

Thank you :)

About [that issue](https://github.com/Ryujinx/Ryujinx/pull/6515#issuecomment-2084422200), it also happens on Linux. It's odd because that issue doesn't  happen on master.

Log file: [Ryujinx_1.1.0+38c2cf8_2024-05-03_21-55-31.log](https://github.com/Ryujinx/Ryujinx/files/15206226/Ryujinx_1.1.0%2B38c2cf8_2024-05-03_21-55-31.log)

TSRBerry commented:
As the exception says, the file is used by another process. Make sure there is nothing else accessing the file and try again.
Most of the time this happens, people open Ryujinx twice.

frafes commented:
Unfortunately I couldn't find what is causing this. Anyway, if I find something new, I'll let you know. Thank you!

AcK77 commented on src/Ryujinx.Gtk3/UI/MainWindow.cs:
```diff
@@ -783,7 +789,7 @@ public void PerformanceCheck()
         }
 
-        private bool LoadApplication(string path, bool isFirmwareTitle)
+        private bool LoadApplication(string path, ulong titleId, bool isFirmwareTitle)
```
```suggestion
        private bool LoadApplication(string path, ulong applicationId, bool isFirmwareTitle)
```

AcK77 commented on src/Ryujinx.Gtk3/UI/Widgets/GameTableContextMenu.cs:
```diff
@@ -36,17 +37,13 @@ public partial class GameTableContextMenu : Menu
-        private readonly string _titleName;
-        private readonly string _titleIdText;
-        private readonly ulong _titleId;
+        private readonly ApplicationData _title;
```
```suggestion
        private readonly ApplicationData _applicationData;
```

AcK77 commented on src/Ryujinx.Gtk3/UI/Windows/DlcWindow.cs:
```diff
@@ -32,16 +36,16 @@ public class DlcWindow : Window
+        public DlcWindow(VirtualFileSystem virtualFileSystem, string titleId, ApplicationData title) : this(new Builder("Ryujinx.Gtk3.UI.Windows.DlcWindow.glade"), virtualFileSystem, titleId, title) { }
 
-        private DlcWindow(Builder builder, VirtualFileSystem virtualFileSystem, string titleId, string titleName) : base(builder.GetRawOwnedObject("_dlcWindow"))
+        private DlcWindow(Builder builder, VirtualFileSystem virtualFileSystem, string applicationId, ApplicationData title) : base(builder.GetRawOwnedObject("_dlcWindow"))
```
```suggestion
        private DlcWindow(Builder builder, VirtualFileSystem virtualFileSystem, string applicationId, ApplicationData applicationData) : base(builder.GetRawOwnedObject("_dlcWindow"))
```

AcK77 commented on src/Ryujinx.Gtk3/UI/Windows/TitleUpdateWindow.cs:
```diff
@@ -24,7 +27,7 @@ public class TitleUpdateWindow : Window
         private readonly MainWindow _parent;
         private readonly VirtualFileSystem _virtualFileSystem;
-        private readonly string _titleId;
+        private readonly ApplicationData _title;
```
```suggestion
        private readonly ApplicationData _applicationData;
```

AcK77 commented on src/Ryujinx.HLE/Loaders/Processes/Extensions/NcaExtensions.cs:
```diff
@@ -7,16 +7,25 @@
-    static class NcaExtensions
+    public static class NcaExtensions
     {
+        private static readonly TitleUpdateMetadataJsonSerializerContext _titleSerializerContext = new(JsonHelper.GetDefaultSerializerOptions());
```
Should be `ApplicationUpdate...` and `_applicationSerializer...`

TSRBerry replied:
I'll address the class name in a separate PR.

AcK77 commented on src/Ryujinx.HLE/Loaders/Processes/Extensions/PartitionFileSystemExtensions.cs:
```diff
@@ -1,26 +1,58 @@
+            return programs;
+        }
+
+        internal static (bool, ProcessResult) TryLoad<TMetaData, TFormat, THeader, TEntry>(this PartitionFileSystemCore<TMetaData, TFormat, THeader, TEntry> partitionFileSystem, Switch device, string path, ulong titleId, out string errorMessage)
```
```suggestion
        internal static (bool, ProcessResult) TryLoad<TMetaData, TFormat, THeader, TEntry>(this PartitionFileSystemCore<TMetaData, TFormat, THeader, TEntry> partitionFileSystem, Switch device, string path, ulong applicationId, out string errorMessage)
```

AcK77 commented on src/Ryujinx.HLE/Loaders/Processes/ProcessLoader.cs:
```diff
@@ -32,7 +32,7 @@ public ProcessLoader(Switch device)
         }
 
-        public bool LoadXci(string path)
+        public bool LoadXci(string path, ulong titleId)
```
```suggestion
        public bool LoadXci(string path, ulong applicationId)
```

AcK77 commented on src/Ryujinx.HLE/Loaders/Processes/ProcessLoader.cs:
```diff
@@ -66,13 +66,13 @@ public bool LoadXci(string path)
         }
 
-        public bool LoadNsp(string path)
+        public bool LoadNsp(string path, ulong titleId)
```
```suggestion
        public bool LoadNsp(string path, ulong applicationId)
```

AcK77 commented on src/Ryujinx.HLE/Switch.cs:
```diff
@@ -73,19 +73,19 @@ public bool LoadCart(string exeFsDir, string romFsFile = null)
         }
 
-        public bool LoadXci(string xciFile)
+        public bool LoadXci(string xciFile, ulong titleId = 0)
```
```suggestion
        public bool LoadXci(string xciFile, ulong applicationId = 0)
```

AcK77 commented on src/Ryujinx.HLE/Switch.cs:
```diff
@@ -73,19 +73,19 @@ public bool LoadCart(string exeFsDir, string romFsFile = null)
         }
 
-        public bool LoadNsp(string nspFile)
+        public bool LoadNsp(string nspFile, ulong titleId = 0)
```
```suggestion
        public bool LoadNsp(string nspFile, ulong applicationId = 0)
```

AcK77 commented on src/Ryujinx.UI.Common/App/ApplicationLibrary.cs:
```diff
@@ -70,272 +73,240 @@ private static byte[] GetResourceBytes(string resourceName)
+                            break;
                         }
-                        else if (extension == ".nro")
+                    case ".nsp":
+                    case ".pfs0":
```
You should add `{}` since you added them on other cases

AcK77 commented on src/Ryujinx.UI.Common/App/ApplicationLibrary.cs:
```diff
@@ -554,10 +631,29 @@ public static ApplicationMetadata LoadAndSaveMetaData(string titleId, Action<App
         }
 
-        public byte[] GetApplicationIcon(string applicationPath, Language desiredTitleLanguage)
+        public byte[] GetApplicationIcon(string applicationPath, Language desiredTitleLanguage, ulong titleId)
```
```suggestion
        public byte[] GetApplicationIcon(string applicationPath, Language desiredTitleLanguage, ulong applicationId)
```

AcK77 commented on src/Ryujinx/Assets/Locales/en_US.json:
```diff
@@ -643,6 +644,8 @@
   "OpenSetupGuideMessage": "Open the Setup Guide",
   "NoUpdate": "No Update",
   "TitleUpdateVersionLabel": "Version {0}",
+  "TitleBundledUpdateVersionLabel": "Bundled: Version {0}",
+  "TitleBundledDlcLabel": "Bundled:",
```
`Title` > `Application`

TSRBerry replied:
Probably better to address with the other labels in a new PR.

AcK77 reviewed (changes requested):
You should double check if `Title` is replaced by `Application`. I know there are a bunch of `TitleUpdate` even in windows. Maybe that could be renamed later here because it's out of scope here.

TSRBerry commented:
Let's do that in a separate PR, since this one is already large enough and I would need to change more strings than currently affected by this PR.

riperiperi commented on src/Ryujinx.Gtk3/UI/Widgets/GameTableContextMenu.cs:
```diff
@@ -55,23 +52,13 @@ public GameTableContextMenu(MainWindow parent, VirtualFileSystem virtualFileSyst
-            _openSaveUserDirMenuItem.Sensitive = !Utilities.IsZeros(controlData.ByteSpan) && controlData.Value.UserAccountSaveDataSize > 0;
-            _openSaveDeviceDirMenuItem.Sensitive = !Utilities.IsZeros(controlData.ByteSpan) && controlData.Value.DeviceSaveDataSize > 0;
-            _openSaveBcatDirMenuItem.Sensitive = !Utilities.IsZeros(controlData.ByteSpan) && controlData.Value.BcatDeliveryCacheStorageSize > 0;
+            _openSaveUserDirMenuItem.Sensitive = !Utilities.IsZeros(_applicationData.ControlHolder.ByteSpan) && _applicationData.ControlHolder.Value.UserAccountSaveDataSize > 0;
```
Maybe `Utilities.IsZeros(_applicationData.ControlHolder.ByteSpan)` should only be checked once.

riperiperi commented on src/Ryujinx/UI/Models/DownloadableContentModel.cs:
```diff
@@ -24,6 +25,9 @@ public bool Enabled
         public string FileName => Path.GetFileName(ContainerPath);
 
+        public string Label =>
+            Path.GetExtension(FileName)?.ToLower() == ".xci" ? $"{LocaleManager.Instance[LocaleKeys.TitleBundledDlcLabel]} {FileName}" : FileName;
```
Duplicating the xci extension check around the place is a bit messy, though I'm not sure where to put it otherwise.

TSRBerry replied:
Me neither, that's how it ended up there. I think it's fine to leave it like that though.

riperiperi commented on src/Ryujinx/UI/ViewModels/MainWindowViewModel.cs:
```diff
@@ -1494,7 +1493,15 @@ public async Task OpenFile()
+                if (ApplicationLibrary.TryGetApplicationsFromFile(result[0].Path.LocalPath,
+                        out List<ApplicationData> applications))
+                {
+                    await LoadApplication(applications[0]);
```
I usually launch applications via the file picker. It would be useful if there were some additional dialog that shows up when there are multiple applications, but that should be separate from this PR.

riperiperi reviewed (approved):
Looks good, seems to work fine in normal games, and multi-program ones like 3D all-stars, though I don't have any multi-game XCIs to test.

gdkchan reviewed (approved):
lgtm, thanks. I tested a few games, on different formats and with/without updates, DLCs and mods, and found no issues. I don't have a multiple game XCI on hand to test right now.

Littl3Guy commented:
Super Mario 3D All Stars doesn't seem to apply updates when launching any of the games
![unknown_2024 07 16-19 18](https://github.com/user-attachments/assets/131a3ab2-266a-4c96-bc89-27fe164e5ed6)

Littl3Guy commented:
> Super Mario 3D All Stars doesn't seem to apply updates when launching any of the games ![unknown_2024 07 16-19 18](https://private-user-images.githubusercontent.com/115218507/349300647-131a3ab2-266a-4c96-bc89-27fe164e5ed6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExNzMzODYsIm5iZiI6MTcyMTE3MzA4NiwicGF0aCI6Ii8xMTUyMTg1MDcvMzQ5MzAwNjQ3LTEzMWEzYWIyLTI2NmEtNGM5Ni1iYzg5LTI3ZmUxNjRlNWVkNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxNlQyMzM4MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYTI1Mzg0OGUwN2RhNTE4MTU1M2M4NTNmYWU0ODI5NjM3ZTI0NmJlZDJkMzk2ZDI4OTlmMDBlYzUxZWNhMmI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.VLW4en6q7GjgXy9AJH6WWAH04NExsgX4Oc_TMRG0kPE)

This is actually a bad example since Galaxy launches 1.0.0 before this, but Sunshine is launching 1.0.0 when it should be 1.1.0

Fafner1000 commented:
Maybe I am understanding wrong the scope of this PR but I have sereval games with updates and dlc already applied to xci file. Zelda Breath of the Wild, Xenoblade 2, Xenoblade 3., Marvel Ultimate Alliance 3, etc. Yuzu recognizes all updates and DLC applied but ryujinx still doesnt, games still launch as version 1.0.0 with no DLC.

riperiperi commented:
Updates aren't meant to be bundled into the XCI file. Apply them correctly from their separate NSP.

TSRBerry commented:
> > Super Mario 3D All Stars doesn't seem to apply updates when launching any of the games ![unknown_2024 07 16-19 18](https://private-user-images.githubusercontent.com/115218507/349300647-131a3ab2-266a-4c96-bc89-27fe164e5ed6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjExNzMzODYsIm5iZiI6MTcyMTE3MzA4NiwicGF0aCI6Ii8xMTUyMTg1MDcvMzQ5MzAwNjQ3LTEzMWEzYWIyLTI2NmEtNGM5Ni1iYzg5LTI3ZmUxNjRlNWVkNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxNlQyMzM4MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYTI1Mzg0OGUwN2RhNTE4MTU1M2M4NTNmYWU0ODI5NjM3ZTI0NmJlZDJkMzk2ZDI4OTlmMDBlYzUxZWNhMmI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.VLW4en6q7GjgXy9AJH6WWAH04NExsgX4Oc_TMRG0kPE)
> 
> This is actually a bad example since Galaxy launches 1.0.0 before this, but Sunshine is launching 1.0.0 when it should be 1.1.0
> 
> Before: ![unknown_2024 07 16-20 28](https://private-user-images.githubusercontent.com/115218507/349312790-eb12171a-92e7-45c0-99a5-a34219f4fab2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzMjUwMzIsIm5iZiI6MTcyMTMyNDczMiwicGF0aCI6Ii8xMTUyMTg1MDcvMzQ5MzEyNzkwLWViMTIxNzFhLTkyZTctNDVjMC05OWE1LWEzNDIxOWY0ZmFiMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOFQxNzQ1MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zZDE0ZDMzODg3OGExZTY5MmY3ZGNkYzdkMDJmNTAzMDU5ZTNlZmVhOTkxYzBkNzlkOWQ0YTJlZGY4YzI0ZDgwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.eheD89IAuatsSdj3GX01VCCJ2j46TkCx33RVCm1E62I)
> 
> After: ![unknown_2024 07 16-20 27](https://private-user-images.githubusercontent.com/115218507/349312811-a2ca6cb9-2a9b-4611-8a66-6cb323c122e1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzMjUwMzIsIm5iZiI6MTcyMTMyNDczMiwicGF0aCI6Ii8xMTUyMTg1MDcvMzQ5MzEyODExLWEyY2E2Y2I5LTJhOWItNDYxMS04YTY2LTZjYjMyM2MxMjJlMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzE4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxOFQxNzQ1MzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMDhlZmU2MTgyZGZiY2QyNmUxYzczNTliMDZlMzgxNTUxNDJhMjcwNzYyZmZmM2RhMjMwODAyNGZkYzdjNjhmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.hCSndkICjIjGbZC9yCIbseRp7YJ5lcV7oo9ZsOW3jGA)

@Littl3Guy could you open an issue and upload logs from 1.1.1349 and 1.1.1350+ so I can investigate this?

Littl3Guy commented:
@TSRBerry Piplup already has an issue open https://github.com/Ryujinx/Ryujinx/issues/7044

TSRBerry commented:
I see! However the logfiles seem to be incorrect, could you still upload them for both versions?

Foul-Tarnished commented:
Seems to cause crash with bundled updates (made with yanu) see #7074

TSRBerry commented:
> Id like to add that i put all three turok games into one nsp using nscbuilder and before this PR was merged only turok 1 would show up but now only turok 2 shows up im not sure if this was supposed to solve that or not but the games have simply been backed together into one nsp unmodified so this should show all the games right?

NSPs are not supposed to contain multiple applications. We are logging warnings when this happens and only load the first one.
